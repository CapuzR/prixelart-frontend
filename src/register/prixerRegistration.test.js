import React from 'react';
import PrixerRegistration from "./prixerRegistration";
import {render, fireEvent, waitFor, screen} from '@testing-library/react';
import axiosMock from "axios";
import Select from '@material-ui/core/Select';
jest.mock("axios");

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

beforeEach(()=> {
  render(<PrixerRegistration />);
});

// it('<PrixerRegistration /> renders correctly', () => {
//   const tree = renderer
//     .create(<PrixerRegistration />)
//     .toJSON();
//   expect(tree).toMatchSnapshot();
// });

describe('with invalid input values', () => {
  it('display required errors for specialty, instagram, phone, country, city and T&C when empty', async()=> {
    const specialty = screen.getByText(/Especialidad/i);
    const instagram = screen.getByLabelText(/Instagram/i);
    const dateOfBirth = screen.getByLabelText(/Fecha de Nacimiento/i);
    const phone = screen.getByLabelText(/Teléfono/i);
    const country = screen.getByLabelText(/País/i);
    const city = screen.getByLabelText(/Ciudad/i);
    // const password = screen.getByLabelText(/Acepto/i);
    const prixerRegister = screen.getByText(/Guardar e ir a mi perfil/i);

    expect(specialty.textContent).toBe("Especialidad");
    expect(instagram.textContent).toBe("");
    expect(dateOfBirth.textContent).toBe("");
    expect(phone.textContent).toBe("");
    expect(country.textContent).toBe("");
    expect(city.textContent).toBe("");

    await waitFor(async()=> {
      fireEvent.submit(prixerRegister);
    });

    expect(screen.getByText('Por favor completa todos los campos requeridos.')).toBeTruthy();
  });
});

describe('with valid input values', () => {
  it('Inputting form login states works correctly', ()=> {
    const mockCallback = jest.fn();
    const { getByTestId } = render(
      <div>
        <Select
          native={true}
          onChange={mockCallback}
          data-testid="select"
          defaultValue="Ambas"
        >
          <option value="Ambas">Ambas</option>
          <option value="fotografía">fotografía</option>
          <option value="Diseñador">Diseñador</option>
        </Select>
      </div>
    );

    const specialty = getByTestId("select");
    const instagram = screen.getByLabelText(/Instagram/i);
    const dateOfBirth = screen.getByLabelText(/Fecha de Nacimiento/i);
    const phone = screen.getByLabelText(/Teléfono/i);
    const country = screen.getByLabelText(/País/i);
    const city = screen.getByLabelText(/Ciudad/i);

    fireEvent.click(specialty.childNodes[0].childNodes[0], {target: {value: 'fotografía', textContent: 'fotografía'}});
    fireEvent.change(instagram, {target: {value: '@capuzr', textContent: '@capuzr'}});
    fireEvent.change(dateOfBirth, {target: {value: '12/05/1991', textContent: '12/05/1991'}});
    fireEvent.change(phone, {target: {value: '+584143201028', textContent: '+584143201028'}});
    fireEvent.change(country, {target: {value: 'Venezuela', textContent: 'Venezuela'}});
    fireEvent.change(city, {target: {value: 'Caracas', textContent: 'Caracas'}});

    expect(specialty.childNodes[0].childNodes[0].textContent).toBe('fotografía');
    expect(instagram.textContent).toBe('@capuzr');
    expect(dateOfBirth.textContent).toBe('12/05/1991');
    expect(phone.textContent).toBe('+584143201028');
    expect(country.textContent).toBe('Venezuela');
    expect(city.textContent).toBe('Caracas');
  });
  
  it('calls api with data from form with all fields', async()=> {
    const mockCallback = jest.fn();
    const { getByTestId } = render(
      <div>
        <Select
          native={true}
          onChange={mockCallback}
          data-testid="select"
          defaultValue="Ambas"
        >
          <option value="Ambas">Ambas</option>
          <option value="fotografía">fotografía</option>
          <option value="Diseñador">Diseñador</option>
        </Select>
      </div>
    );

    const specialty = getByTestId("select");
    const instagram = screen.getByLabelText(/Instagram/i);
    const dateOfBirth = screen.getByLabelText(/Fecha de Nacimiento/i);
    const phone = screen.getByLabelText(/Teléfono/i);
    const country = screen.getByLabelText(/País/i);
    const city = screen.getByLabelText(/Ciudad/i);
    const prixerRegister = screen.getByText(/Guardar e ir a mi perfil/i);
    const axMock = axiosMock.post.mockResolvedValue({data: {success: true}});

    await waitFor(async()=> {
      fireEvent.click(specialty.childNodes[0].childNodes[0], {target: {value: 'Fotógrafía', textContent: 'Fotógrafía'}});
      fireEvent.change(instagram, {target: {value: '@capuzr', textContent: '@capuzr'}});
      fireEvent.change(dateOfBirth, {target: {value: '12/05/1991', textContent: '12/05/1991'}});
      fireEvent.change(phone, {target: {value: '+584143201028', textContent: '+584143201028'}});
      fireEvent.change(country, {target: {value: 'Venezuela', textContent: 'Venezuela'}});
      fireEvent.change(city, {target: {value: 'Caracas', textContent: 'Caracas'}});
    });

    await waitFor(async()=> {
      fireEvent.submit(prixerRegister);
    });

    expect(axMock).toHaveBeenCalled();
  });
  
  it('username already has a Prixer', async()=> {
  const mockCallback = jest.fn();
  const { getByTestId } = render(
    <div>
      <Select
        native={true}
        onChange={mockCallback}
        data-testid="select"
        defaultValue="Ambas"
      >
        <option value="Ambas">Ambas</option>
        <option value="fotografía">fotografía</option>
        <option value="Diseñador">Diseñador</option>
      </Select>
    </div>
  );

  const specialty = getByTestId("select");
  const instagram = screen.getByLabelText(/Instagram/i);
  const dateOfBirth = screen.getByLabelText(/Fecha de Nacimiento/i);
  const phone = screen.getByLabelText(/Teléfono/i);
  const country = screen.getByLabelText(/País/i);
  const city = screen.getByLabelText(/Ciudad/i);
  const prixerRegister = screen.getByText(/Guardar e ir a mi perfil/i);
  axiosMock.post.mockResolvedValue({data: {
    success: false,
    message: 'Disculpa, este usuario ya está asignado a un Prixer'
  }});

  await waitFor(async()=> {
    fireEvent.click(specialty.childNodes[0].childNodes[0], {target: {value: 'fotografía', textContent: 'fotografía'}});
    fireEvent.change(instagram, {target: {value: '@capuzr', textContent: '@capuzr'}});
    fireEvent.change(dateOfBirth, {target: {value: '12/05/1991', textContent: '12/05/1991'}});
    fireEvent.change(phone, {target: {value: '+584143201028', textContent: '+584143201028'}});
    fireEvent.change(country, {target: {value: 'Venezuela', textContent: 'Venezuela'}});
    fireEvent.change(city, {target: {value: 'Caracas', textContent: 'Caracas'}});
  });

  await waitFor(async()=> {
    fireEvent.submit(prixerRegister);
  });
    expect(screen.getByText('Disculpa, este usuario ya está asignado a un Prixer')).toBeTruthy();
  });
  
  it('redirects to prixer Profile', async()=> {
    const mockCallback = jest.fn();
    const { getByTestId } = render(
      <div>
        <Select
          native={true}
          onChange={mockCallback}
          data-testid="select"
          defaultValue="Ambas"
        >
          <option value="Ambas">Ambas</option>
          <option value="fotografía">fotografía</option>
          <option value="Diseñador">Diseñador</option>
        </Select>
      </div>
    );
  
    const specialty = getByTestId("select");
    const instagram = screen.getByLabelText(/Instagram/i);
    const dateOfBirth = screen.getByLabelText(/Fecha de Nacimiento/i);
    const phone = screen.getByLabelText(/Teléfono/i);
    const country = screen.getByLabelText(/País/i);
    const city = screen.getByLabelText(/Ciudad/i);
    const prixerRegister = screen.getByText(/Guardar e ir a mi perfil/i);
    axiosMock.post.mockResolvedValue({data: {
      success: true
    }});
  
    await waitFor(async()=> {
      fireEvent.click(specialty.childNodes[0].childNodes[0], {target: {value: 'fotografía', textContent: 'fotografía'}});
      fireEvent.change(instagram, {target: {value: '@capuzr', textContent: '@capuzr'}});
      fireEvent.change(dateOfBirth, {target: {value: '12/05/1991', textContent: '12/05/1991'}});
      fireEvent.change(phone, {target: {value: '+584143201028', textContent: '+584143201028'}});
      fireEvent.change(country, {target: {value: 'Venezuela', textContent: 'Venezuela'}});
      fireEvent.change(city, {target: {value: 'Caracas', textContent: 'Caracas'}});
    });
  
    await waitFor(async()=> {
      fireEvent.submit(prixerRegister);
    });

    expect(mockHistoryPush).toHaveBeenCalledWith({pathname:"/prixer-profile"});
  });
  
  it('renders snackbar with success message', async()=> {
    const mockCallback = jest.fn();
    const { getByTestId } = render(
      <div>
        <Select
          native={true}
          onChange={mockCallback}
          data-testid="select"
          defaultValue="Ambas"
        >
          <option value="Ambas">Ambas</option>
          <option value="fotografía">fotografía</option>
          <option value="Diseñador">Diseñador</option>
        </Select>
      </div>
    );
  
    const specialty = getByTestId("select");
    const instagram = screen.getByLabelText(/Instagram/i);
    const dateOfBirth = screen.getByLabelText(/Fecha de Nacimiento/i);
    const phone = screen.getByLabelText(/Teléfono/i);
    const country = screen.getByLabelText(/País/i);
    const city = screen.getByLabelText(/Ciudad/i);
    const prixerRegister = screen.getByText(/Guardar e ir a mi perfil/i);
    axiosMock.post.mockResolvedValue({data: {
      success: true
    }});
  
    await waitFor(async()=> {
      fireEvent.click(specialty.childNodes[0].childNodes[0], {target: {value: 'fotografía', textContent: 'fotografía'}});
      fireEvent.change(instagram, {target: {value: '@capuzr', textContent: '@capuzr'}});
      fireEvent.change(dateOfBirth, {target: {value: '12/05/1991', textContent: '12/05/1991'}});
      fireEvent.change(phone, {target: {value: '+584143201028', textContent: '+584143201028'}});
      fireEvent.change(country, {target: {value: 'Venezuela', textContent: 'Venezuela'}});
      fireEvent.change(city, {target: {value: 'Caracas', textContent: 'Caracas'}});
    });
  
    await waitFor(async()=> {
      fireEvent.submit(prixerRegister);
    });
    
    expect(screen.queryByText('Registro de Prixer exitoso.')).toBeInTheDocument();
  });
});