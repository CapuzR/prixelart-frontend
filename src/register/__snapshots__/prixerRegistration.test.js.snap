// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PrixerRegistration /> renders correctly 1`] = `
<main
  className="MuiContainer-root MuiContainer-maxWidthXs"
>
  <div
    className="makeStyles-paper-1"
  >
    <div
      className="MuiAvatar-root MuiAvatar-circle makeStyles-avatar-2 MuiAvatar-colorDefault"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
        />
      </svg>
    </div>
    <h1
      className="MuiTypography-root MuiTypography-h5"
      onChange={[Function]}
    >
      Comparte con tus futuros clientes
    </h1>
    <form
      className="makeStyles-form-3"
      noValidate={true}
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-animated"
            id="demo-simple-select-label"
          >
            Especialidad
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline"
            onClick={[Function]}
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="specialty specialty"
              className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              id="specialty"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </div>
            <input
              aria-hidden={true}
              className="MuiSelect-nativeInput"
              onAnimationStart={[Function]}
              onChange={[Function]}
              tabIndex={-1}
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9 MuiGrid-grid-sm-9"
        >
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink={false}
              htmlFor="instagram"
              id="instagram-label"
            >
              Instagram
              <span
                aria-hidden={true}
                className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="fname"
                autoFocus={true}
                className="MuiInputBase-input MuiOutlinedInput-input"
                disabled={false}
                id="instagram"
                name="instagram"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="text"
              />
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="PrivateNotchedOutline-legendLabelled-7"
                >
                  <span>
                    Instagram
                     *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-sm-6"
        >
          <div
            className="MuiFormControl-root MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
              data-shrink={true}
              htmlFor="dateOfBirth"
              id="dateOfBirth-label"
            >
              Fecha de Nacimiento
            </label>
            <div
              className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiInput-input"
                defaultValue="2016-07-06"
                disabled={false}
                id="dateOfBirth"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="date"
              />
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink={false}
              htmlFor="phone"
              id="phone-label"
            >
              Teléfono
              <span
                aria-hidden={true}
                className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="phone"
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input"
                disabled={false}
                id="phone"
                name="phone"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="text"
              />
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="PrivateNotchedOutline-legendLabelled-7"
                >
                  <span>
                    Teléfono
                     *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink={false}
              htmlFor="country"
              id="country-label"
            >
              País
              <span
                aria-hidden={true}
                className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="current-password"
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input"
                disabled={false}
                id="country"
                name="country"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="country"
              />
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="PrivateNotchedOutline-legendLabelled-7"
                >
                  <span>
                    País
                     *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
        >
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink={false}
              htmlFor="city"
              id="city-label"
            >
              Ciudad
              <span
                aria-hidden={true}
                className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoComplete="current-password"
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input"
                disabled={false}
                id="city"
                name="city"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="city"
              />
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="PrivateNotchedOutline-legendLabelled-7"
                >
                  <span>
                    Ciudad
                     *
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
      >
        <label
          className="MuiFormControlLabel-root"
        >
          <span
            aria-disabled={false}
            className="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-9 MuiCheckbox-root MuiCheckbox-colorPrimary MuiIconButton-colorPrimary"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={null}
          >
            <span
              className="MuiIconButton-label"
            >
              <input
                className="PrivateSwitchBase-input-12"
                data-indeterminate={false}
                onChange={[Function]}
                type="checkbox"
                value="allowExtraEmails"
              />
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </span>
          </span>
          <span
            className="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
          >
            Acepto los Términos y condiciones.
          </span>
        </label>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-4 MuiButton-containedPrimary MuiButton-fullWidth"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
        value="submit"
      >
        <span
          className="MuiButton-label"
        >
          Guardar e ir a mi perfil
        </span>
      </button>
    </form>
  </div>
  <div
    className="MuiBox-root MuiBox-root-13"
  >
    <p
      className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
    >
      Copyright © 
      <a
        className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
        href="https://prixelart.com/"
        onBlur={[Function]}
        onFocus={[Function]}
      >
        prixelart.com
      </a>
       
      2020
      .
    </p>
  </div>
</main>
`;
